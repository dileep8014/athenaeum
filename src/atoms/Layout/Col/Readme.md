The `Col` component is a part of the `Layout` component. See `Layout` component for usage details.
