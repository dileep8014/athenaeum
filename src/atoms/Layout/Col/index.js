import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from '../layout.module.scss';

function Col(props) {
  const {
    children,
    className,
    smallCols,
    mediumCols,
    largeCols,
    xLargeCols,
    xxLargeCols,
    padding,
    fullwidth,
    flex,
    style,
    bottomSpacing
  } = props;

  return (
    <div
      style={style}
      className={classnames(
        styles['col'],
        { [styles[`col-sm-${smallCols}`]]: smallCols },
        { [styles[`col-md-${mediumCols}`]]: mediumCols },
        { [styles[`col-lg-${largeCols}`]]: largeCols },
        { [styles[`col-xlg-${xLargeCols}`]]: xLargeCols },
        { [styles[`col-xxlg-${xxLargeCols}`]]: xxLargeCols },
        { [styles.fullwidth]: fullwidth },
        { [styles.flex]: flex },
        styles[`bottom-spacing-${bottomSpacing}`],
        padding && styles.padding,
        className,
      )}
    >
      { children }
    </div>
  );
}

Col.propTypes = {
  /**
   * Supply any additional class names.
   */
  className: PropTypes.string,

  /**
   * sets bottom spacing between children in Layout
   * - see [Spacer](#spacer) for appropriate values
   */
  bottomSpacing: PropTypes.string,

  /**
   * Overides column grid width within a viewport from `0px` up. Width can also be set to 'auto'.
   */
  smallCols: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.oneOf([ 'auto' ])
  ]),

  /**
   * Sets col grid width in viewport from `768px` up. Width can also be set to 'auto'.
   */
  mediumCols: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.oneOf([ 'auto' ])
  ]),

  /**
   * Sets col grid width in viewport from `1024px` up. Width can also be set to 'auto'.
   */
  largeCols: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.oneOf([ 'auto' ])
  ]),

  /**
   * Sets col grid width in viewport from `1280px` up. Width can also be set to 'auto'.
   */
  xLargeCols: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.oneOf([ 'auto' ])
  ]),
  /**
   * Sets col grid width in viewport from `1440px` up. Width can also be set to 'auto'.
   */
  xxLargeCols: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.oneOf([ 'auto' ])
  ]),

  /**
   * Overides flex, changes Col to grow to fill empty space in the row (`flex-grow: 1`)
   */
  flex: PropTypes.bool,

  /**
   * Overides and removes padding set from Layout
   */
  fullwidth: PropTypes.bool,

  /**
   * Adds standard (~6px) left/right padding to the `Layout`
   */
  padding: PropTypes.bool,

  /**
   * adds additional styles to the column as a React styles object  ``
   * - For useful positional styles, [Checkout this Guide to Flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/)
   */
  style: PropTypes.object
};


export default Col;
