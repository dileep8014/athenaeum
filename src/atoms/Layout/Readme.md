Layout is broken out into a 12 column grid of `Col`s.

Shorthand option:
You don't have to explicitly wrap each child in a `Col`. All children of `Layout` will get wrapped in `Col` automatically.
Any layout specific props that are meant for the `Col` will be passed to the `Col` wrapper accordingly and will override any props that may have been given it by the `Layout` component. All child component props will stay with the component itself.


## How to use Layout
```html
  // Bad (don't pass props that don't exist.)
  <Layout smallCols={[ 6 ]}>
    <Button variant='info'>I am button</Button>
    <Button variant='outline' smallCols={3}>some button</Button>
  </Layout>

  // Good
  <Layout smallCols={[ 6 ]}>
    <Button variant='info'>I am button</Button>
    <Button variant='outline'>some button</Button>
  </Layout>

  // Good
  <Layout smallCols={[ 6 ]}>
    <Button variant='info'>I am button</Button>
    <Col smallCols={3}>
      <Button variant='outline'>some button</Button>
    </Col>
  </Layout>
```


## Mixed Layout Examples with Shorthand, Cols, children, and Overrides:

```example
    const exampleStyles = require('./example.module.scss');

    <div className={exampleStyles.example}>
      <Layout
        smallCols={[ 6 ]}
        style={{ justifyContent: 'space-between', alignItems: 'center' }}
      >
        <div />

        <Col smallCols={3}>
          <div style={{ background: 'red', margin: '10px' }}> Some child</div>
        </Col>

        <Col smallCols={2} style={{ alignSelf: 'flex-end' }} />

        <Col smallCols={3}>
          <Button variant='info'>I am button</Button>
        </Col>

        <Button variant='outline'>some button</Button>


        <Col style={{ padding: '30px' }}>
          <Button variant='info'>I am also button</Button>
        </Col>

      </Layout>
    </div>
```

## Small 2 Col, Medium 3 Col Grid Example:


```example
    const exampleStyles = require('./example.module.scss');

    <div className={exampleStyles.example}>
      <Layout
        smallCols={[ 6 ]}
        mediumCols={[ 4 ]}
      >
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
      </Layout>
    </div>
```

## Nested Grid Example:

```example
    const exampleStyles = require('./example.module.scss');

    <div className={exampleStyles.example}>
      <Layout
        smallCols={[ 3 ]}
      >

        <Col smallCols={7}>
          <Layout smallCols={[ 4 ]}>
            <div> col 2 </div>
            <div> col 2 </div>
            <div> col 2 </div>
          </Layout>
        </Col>

        <Col smallCols={5}> col 1 </Col>

        <Col smallCols={5}> col 1 </Col>

        <Col smallCols={7} fullwidth>
          <Layout smallCols={[ 4 ]} fullwidth>
            <div> col 2 </div>
            <div> col 2 </div>
            <div> col 2 </div>
          </Layout>
        </Col>

        <div> col 1 </div>
        <div> col 1 </div>
        <div> col 1 </div>
        <div> col 1 </div>
        <div> col 1 </div>

      </Layout>
    </div>
```
