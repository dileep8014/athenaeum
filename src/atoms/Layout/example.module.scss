.example {
  div {
    background: rgba(80, 98, 255, 0.25);
    min-height: 50px;

    > div {
      outline: 1px dashed rgba(80, 98, 255, 0.25);
    }
  }
}
