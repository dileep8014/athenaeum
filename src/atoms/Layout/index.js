import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from './layout.module.scss';

function Layout( props ) {
  const {
    children,
    childCols,
    breakpoint,
    horizontalAlign,
    verticalAlign
  } = props;

  // start at -1 so can bump to 0 in the conditional below
  let colsIdx = -1;

  const kids = React.Children.map( children, ( child ) => {
    // colsIdx mirrors idx until idx does not exist in childCols
    // this means that there are more children than childCols provided
    // in that case, reset colsIdx to 0 and continue counting up until next reset
    if ( !childCols[++colsIdx] ) {
      colsIdx = 0;
    }

    return (
      <div
        className={classnames(
          // grab childCols value with colsIdx
          // if childCols[idx] is undefined, default to 12 columns
          styles[`column-${childCols[colsIdx] || 12}`]
        )}
      >
        { child }
      </div>
    );
  });

  const hAlign = `h-${horizontalAlign}`;
  const vAlign = `v-${verticalAlign}`;

  return (
    <div>
      <div
        className={
          classnames( styles.layout, styles[breakpoint], styles[hAlign], styles[vAlign] )
        }
      >
        { kids }
      </div>
    </div>
  );
}

Layout.propTypes = {
  /**
   * determines pixel breakpoint for columns and rows
   */
  breakpoint: PropTypes.string,

  /**
   * each number in array sets column width for each child in Layout
   */
  childCols: PropTypes.arrayOf( PropTypes.number ),

  /**
   * determines alignment of columns horizontally
   */
  horizontalAlign: PropTypes.string,

  /**
   * determines alignment of columns vertically
   */
  verticalAlign: PropTypes.string
};

Layout.defaultProps = {
  childCols: [ 12 ],
  breakpoint: 'small',
  horizontalAlign: 'start',
  verticalAlign: 'center'
};

export default Layout;
