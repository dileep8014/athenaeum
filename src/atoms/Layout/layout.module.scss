@import 'src/assets/stylesheets/global';

$gutter-width: .75rem;

@mixin g-unit($cols: 1, $total-col-count: 12) {
  $width: calc( ( #{$cols} / #{$total-col-count} ) * 100%);

  width: $width;
  flex-basis: $width;
  flex-grow: 0;
  flex-shrink: 1;
}

@mixin createCols( $size ) {
  @for $i from 1 through 12 {
    .col-#{$size}-#{$i} {
      @include g-unit($i);
      padding-left: $gutter-width;
      padding-right: $gutter-width;
    }
  }
}

.layout {
  display: flex;
  flex-wrap: wrap;
}

@media #{$small-up} {
  @include createCols( 'sm' );
}

@media #{$medium-up} {
  @include createCols( 'md' );

  .layout {
    padding-left: $gutter-width;
    padding-right: $gutter-width;
  }
}

@media #{$large-up} {
  @include createCols( 'lg' );
}

@media #{$xlarge-up} {
  @include createCols( 'xlg' );
}

@media #{$xxlarge-up} {
  @include createCols( 'xxlg' );
}

// Utility classes must come after base styles
// so that they can override the base in the cascade

.fullwidth {
  padding: 0;
}

.flex {
  flex-grow: 1;
}
