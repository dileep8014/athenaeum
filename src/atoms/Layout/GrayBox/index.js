import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from '../layout.module.scss';
import Col from '../Col';

function GrayBox(props) {
  const {
    className,
  } = props;

  const classes = cx(
    styles['gray-box'],
    className,
  );

  return (
    <Col
      {...props}
      className={classes}
    />
  );
}

GrayBox.rclName = 'Col';

GrayBox.propTypes = {
  /**
   * Supply any additional class names.
   */
  className: PropTypes.string,
};

export default GrayBox;
