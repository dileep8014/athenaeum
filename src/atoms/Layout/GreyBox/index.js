import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from '../layout.module.scss';
import Col from '../Col';

function GreyBox(props) {
  const {
    className,
  } = props;

  const classes = cx(
    styles['grey-box'],
    className,
  );

  return (
    <Col
      {...props}
      className={classes}
    />
  );
}

GreyBox.rclName = 'Col';

GreyBox.propTypes = {
  /**
   * Supply any additional class names.
   */
  className: PropTypes.string,
};

export default GreyBox;
