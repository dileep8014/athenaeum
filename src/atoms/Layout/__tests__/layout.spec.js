import React from 'react';
import { shallow } from 'enzyme';
import { expect } from 'chai';

import Layout from 'atoms/Layout';
import Col from 'atoms/Layout/Col';

describe('<Layout />', () => {

  it('renders', () => {
    const wrapper = shallow(<Layout />);

    expect(wrapper.length).to.equal(1);
  });


  it('renders its children', () => {
    const wrapper = shallow(
      <Layout smallCols={[ 1 ]}>
        <div className='child' />
        <div className='some-child' />
        <div className='child' />
      </Layout>
    );

    expect(wrapper.find('.child')).to.have.length(2);
    expect(wrapper.find('.some-child')).to.have.length(1);

  });

  it('wraps its children in `<Col>`', () => {
    const wrapper = shallow(
      <Layout smallCols={[ 1 ]}>
        <div className='child' />
        <div className='child' />
        <div className='child' />
      </Layout>
    );

    expect(wrapper.find('Col')).to.have.length(3);
  });

  it('passes its `col` props down to its `<Col>` correctly', () => {
    const wrapper = shallow(
      <Layout
        smallCols={[ 1 ]}
        mediumCols={[ 6 ]}
        largeCols={[ 7, 5 ]}
        bottomSpacing='xLarge'
        fullwidth
        padding
      >
        <div className='child' />
        <div className='child' />
      </Layout>
    );

    expect(wrapper.find('Col').first().props().smallCols).to.be.ok;
    expect(wrapper.find('Col').first().props().smallCols).to.equal(1);

    expect(wrapper.find('Col').first().props().mediumCols).to.be.ok;
    expect(wrapper.find('Col').first().props().mediumCols).to.equal(6);

    expect(wrapper.find('Col').first().props().fullwidth).to.be.true;
    expect(wrapper.find('Col').first().props().padding).not.to.be.true;
    expect(wrapper.find('Col').first().props().bottomSpacing).to.equal('xLarge');
  });

  it('wraps its child `<Col>` with column sizes', () => {
    const wrapper = shallow(
      <Layout
        largeCols={[ 7, 5 ]}
      >
        <div className='child' />
        <div className='child' />
      </Layout>
    );

    expect(wrapper.find('Col').get(0).props.largeCols).to.equal(7);
    expect(wrapper.find('Col').get(1).props.largeCols).to.equal(5);
  });

  it('wraps its child `<Col>` with column sizes', () => {
    const wrapper = shallow(
      <Layout
        smallCols={[ 1 ]}
      >
        <div className='child' />

        <Col smallCols={3}>
          <div className='one-off' />
        </Col>

        <div className='child' />

      </Layout>
    );

    expect(wrapper.children().get(0).props.smallCols).to.equal(1);
    expect(wrapper.children().get(1).props.smallCols).to.equal(3);
    expect(wrapper.children().get(2).props.smallCols).to.equal(1);
  });

});
