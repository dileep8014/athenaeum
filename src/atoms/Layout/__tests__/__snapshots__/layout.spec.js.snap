// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> can create variant via props 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<Layout /> doesn't wrap Cols in \`<Col>\` 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<Layout /> passes its \`col\` props down to its \`<Col>\` correctly 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<Layout /> renders correctly 1`] = `
<div
  className="layout"
/>
`;

exports[`<Layout /> renders its children 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<Layout /> wraps its child \`<Col>\` with column sizes 1`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<Layout /> wraps its child \`<Col>\` with column sizes 2`] = `
ShallowWrapper {
  "length": 1,
}
`;

exports[`<Layout /> wraps its children in \`<Col>\` 1`] = `
ShallowWrapper {
  "length": 1,
}
`;
